File: Python/Object Oriented Design/ParkingLot.py
# Design a parking lot using object oriented design
# The statement never states what kind of parking lot...is it a parking lot with many floors?
# Is a parking lot that you have to pay for (toll)

# This parking lot will have one level with limited spots and have some sort of toll


class ParkingSpaces(object):
    def __init__(self, number_of_spaces):
        self.number_of_spaces = number_of_spaces
        self.spaces_used = 0

    def update_spaces(self):
        self.spaces_used += 1

    def get_max_number_of_spaces(self):
        return self.number_of_spaces

    def get_spaces_used(self):
        return self.spaces_used


class Toll(ParkingSpaces):
    def __init__(self, number_of_spaces):
        super(Toll, self).__init__(number_of_spaces)
        ParkingSpaces(number_of_spaces)
        self.toll_amount = 1.50


    def pay_toll(self):
        if ParkingSpaces.get_spaces_used(self) < ParkingSpaces.get_max_number_of_spaces(self):
            ParkingSpaces.update_spaces(self)
        else:
            print "Parking lot is full"

    def get_toll_amount(self):
        return self.toll_amount


class ParkingLot(Toll):
    def __init__(self, number_of_spaces):
        super(ParkingLot, self).__init__(number_of_spaces)
        self.parking = Toll(number_of_spaces)
        #self.number_of_cars_parked = 0
        self.money_made = 0


    def pay_toll(self):
        Toll.pay_toll(self)
        #self.number_of_cars_parked += 1
        #self.money_made = Toll.get_toll_amount(self) * self.number_of_cars_parked
        self.money_made = Toll.get_toll_amount(self) * ParkingSpaces.get_spaces_used(self)

    def get_number_of_cars_parked(self):
        #return self.number_of_cars_parked
        return ParkingSpaces.get_spaces_used(self)

    def get_money_made(self):
        return str(self.money_made) + "$"


# If I wanted to convert this to a multi level parking lot I would create a new class
# and make it a sub class of ParkingLot
# class MultiLevel(ParkingLot)


def main():
    pl = ParkingLot(10)
    pl.pay_toll()
    pl.pay_toll()
    pl.pay_toll()
    pl.pay_toll()
    pl.pay_toll()
    pl.pay_toll()
    pl.pay_toll()
    pl.pay_toll()
    pl.pay_toll()
    pl.pay_toll()
    print pl.get_money_made()
    print

if __name__ in "__main__":
    main()

File: Java/Hired Programming Practice/src/ObjectOrientedDesign/ParkingLot.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

// Design a parking lot using object oriented design
// The statement never states what kind of parking lot...is it a parking lot with many floors?
// Is a parking lot that you have to pay for (toll)

// This parking lot will have three levels with limited spots, 
// specific vehicle implementations/sizes

package ObjectOrientedDesign;
/**
 *
 * @author marcsantiago
 */

public class ParkingLot extends ParkingLevels{
    private final int [][] levels = getLevels();
    private final int carSpace = 1;
    private final int motoSpace = 1;
    private final int busSpace = 4;
    
    private final Car car;
    private final Motocycle moto;
    private final Bus bus;
    
    
    
    public ParkingLot(){
        car = new Car(carSpace);
        moto = new Motocycle(motoSpace);
        bus = new Bus(busSpace);
    }
       
    
    private String findEmptySpot(int spaceNeeded){
        spaceNeeded = spaceNeeded - 1;
        for (int i = 0; i < getNumberOfLevels(); i++) {
            for (int j = 0; j < getSpacesPerLevel(); j++) {
                if(j + spaceNeeded < getSpacesPerLevel()) {
                    if(levels[i][j] == 0 && levels[i][j + spaceNeeded] == 0){
                        return Integer.toString(i) + " " + Integer.toString(j);
                    }
                }
            }
        }
        return null;
    }
    
    public void addCar(){
        String [] indexValues = new String[2];
        int level;
        int space;
        try{
            indexValues = findEmptySpot(car.getSpaceNeeded()).split(" ");
        }
        catch(Exception e){
            System.out.println("Not enough spaces for vehicle");
        }
        
        level = Integer.valueOf(indexValues[0]);
        space = Integer.valueOf(indexValues[1]);
        
        for (int i = 0; i < moto.getSpaceNeeded(); i++) {
            levels[level][space + i] = 1;
        }
                    
    }
    
    public void addMotocycle(){
        String [] indexValues = new String[2];
        int level;
        int space;
        try{
            indexValues = findEmptySpot(moto.getSpaceNeeded()).split(" ");
        }
        catch(Exception e){
            System.out.println("Not enough spaces for vehicle");
        }
        
        level = Integer.valueOf(indexValues[0]);
        space = Integer.valueOf(indexValues[1]);
        
        for (int i = 0; i < moto.getSpaceNeeded(); i++) {
            levels[level][space + i] = 1;
        }
    }
    
    public void addBus(){
        String [] indexValues = new String[2];
        int level;
        int space;
        try{
            indexValues = findEmptySpot(bus.getSpaceNeeded()).split(" ");
        }
        catch(Exception e){
            System.out.println("Not enough spaces for vehicle");
        }
        
        level = Integer.valueOf(indexValues[0]);
        space = Integer.valueOf(indexValues[1]);
        
        for (int i = 0; i < bus.getSpaceNeeded(); i++) {
            levels[level][space + i] = 1;
        }
        
    }
}

class ParkingLevels{    
    private final int spacesPerLevel = 10;
    private final int numberOfLevels = 3;
    private final int [][] levels = new int[numberOfLevels][spacesPerLevel];
    
    public ParkingLevels(){
        
    }

    public int getNumberOfLevels() {
        return numberOfLevels;
    }

    public int getSpacesPerLevel() {
        return spacesPerLevel;
    }

    public int[][] getLevels() {
        return levels;
    }
    
    public boolean isLevelEmpty(int level){
        return levels[level - 1][0] == 0;
    }
    
    public boolean isLevelFull(int level){
        int spacesUsed = 0;
        for (int i = 0; i < spacesPerLevel; i++) {
            if(levels[level - 1][i] == 1){
                spacesUsed++;
            }
        }
        return spacesUsed == spacesPerLevel;
    } 
}

class Vehicle{
    private final int spacesUsed;
    public Vehicle(int spacesUsed){
        this.spacesUsed = spacesUsed;
    }
    
    public int getSpaceNeeded(){
        return spacesUsed;
    }
}

class Car extends Vehicle{
    public Car(int spacesUsed) {
        super(spacesUsed);
    }
}

class Bus extends Vehicle{
    public Bus(int spacesUsed){
        super(spacesUsed);
    }
}

class Motocycle extends Vehicle{
    public Motocycle(int spacesUsed){
        super(spacesUsed);
    }
}